contracts/governance/Vesting/fouryear/FourYearVestingLogic.sol

```diff
@@ -0,0 +1,305 @@
+	 * @param receiver The receiving address.
+	 * @dev Can be called only by owner.
+	 * */
+	function governanceWithdrawTokens(address receiver) public {
+		require(msg.sender == address(staking), "unauthorized");
+		_withdrawTokens(receiver, true);
+	}
+
+	/**
+	 * @notice Withdraws unlocked tokens from the staking contract and
+	 * forwards them to an address specified by the token owner.
+	 * @param receiver The receiving address.
+	 * */
+	function withdrawTokens(address receiver) public onlyTokenOwner {
+		_withdrawTokens(receiver, false);
+	}
+
+	/**
+	 * @notice Collect dividends from fee sharing proxy.
+	 * @param _loanPoolToken The loan pool token address.
+	 * @param _maxCheckpoints Maximum number of checkpoints to be processed.
+	 * @param _receiver The receiver of tokens or msg.sender
+	 * */
+	function collectDividends(
+		address _loanPoolToken,
+		uint32 _maxCheckpoints,
+		address _receiver
+	) public onlyTokenOwner {
+		require(_receiver != address(0), "receiver address invalid");
+
+		/// @dev Invokes the fee sharing proxy.
+		feeSharingProxy.withdraw(_loanPoolToken, _maxCheckpoints, _receiver);
+
+		emit DividendsCollected(msg.sender, _loanPoolToken, _receiver, _maxCheckpoints);
+	}
+
+	/**
+	 * @notice Allows the owners to migrate the positions
+	 * to a new staking contract.
+	 * */
+	function migrateToNewStakingContract() public onlyOwners {
+		staking.migrateToNewStakingContract();
+		staking = Staking(staking.newStakingContract());
+		emit MigratedToNewStakingContract(msg.sender, address(staking));
+	}
+
+	/**
+	 * @notice Extends first year stakes for four year vesting contracts.
+	 * @dev Tokens are vested for 4 years. Since the max staking
+	 * period is 3 years and the tokens are unlocked only after the first year is
+	 * passed, hence, we extend the duration of staking for all unlocked tokens for the first
+	 * year by 3 years.
+	 * */
+	function extendStaking() public {
+		uint256 oneYear = startDate.add(52 weeks);
+		uint256[] memory dates;
+		uint96[] memory stakes;
+		(dates, stakes) = staking.getStakes(address(this));
+
+		for (uint256 i = 0; i < dates.length; i++) {
+			if ((dates[i] < block.timestamp) && (dates[i] <= oneYear) && (stakes[i] > 0)) {
+				staking.extendStakingDuration(dates[i], dates[i].add(156 weeks));
+			}
+		}
+	}
+
+	/**
+	 * @notice Stakes tokens according to the vesting schedule. Low level function.
+	 * @dev Once here the allowance of tokens is taken for granted.
+	 * @param _sender The sender of tokens to stake.
+	 * @param _amount The amount of tokens to stake.
+	 * @param _restartStakeSchedule The time from which staking schedule restarts.
+	 * The issue is that we can only stake tokens for a max duration. Thus, we need to restart
+	 * from the lastSchedule.
+	 * @return lastSchedule The max duration for which tokens were staked.
+	 * @return remainingAmount The amount outstanding - to be staked.
+	 * */
+	function _stakeTokens(
```

Comment:
this looks way to complicated and potentially incorrect. i would like to discuss it on a call.

Reply:
1. Modified the code to strictly run for four year vesting(Cliff = 4 weeks, Duration = 156 weeks)
2. Added necessary comments and differentiated cliff from Interval(FOUR_WEEKS)
3. Changed maxDuration to maxInterval to remove confusion with Staking duration. Considered maxInterval = 18*4 weeks for comments. The gas usage amounts to 5.2M approximately when scheduling stakes for maxInterval.
4. Removed test scenarios for other cliff/duration and optimised those.